[Bug 619284] Create a "draft" status for unfinished translations

safwanrahman opened the pull request:
A long waiting patch. :smiley: :sunglasses: Saving a revision as draft is badly needed from localization end. The saving and restoring system can be better by using javascript, but I dont have much Idea about javascript. so leaving it for others to work in future. The javascript in this patch may not be as good because of that, so please consider reviewing. Though most of the work is done in the backend.
@mythmon r?

mythmon commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,14 @@
+    var doc_data = $('#doc_form').serializeArray();
+    var rev_data = $('#rev_form').serializeArray();
+    var total_data = both_data.concat(doc_data, rev_data);
+    $.post( url, total_data );
```
This should give feedback to users, reporting whether the action was a success or failure.

safwanrahman replied:
@mythmon Actually, I dont have much knowledge about javascript. If you can show me any code which do this type of work, I think I can understand how to show a feedback to the users.

mythmon commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,14 @@
     });
   }
 
+  var draftbutton = document.getElementsByClassName('btn-draft')[0];
```
Vanilla-JS DOM access is better written as `var draftButton = document.querySelector('.btn-draft');`

safwanrahman replied:
Thanks. I did not have idea about it

mythmon commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,14 @@
+    var both_data = $('#both_form').serializeArray();
+    var doc_data = $('#doc_form').serializeArray();
+    var rev_data = $('#rev_form').serializeArray();
+    var total_data = both_data.concat(doc_data, rev_data);
```
Mixing Vanilla-JS DOM access and jQuery DOM access is confusing. You should pick one and use that for all the operations.

mythmon commented on kitsune/wiki/forms.py:
```diff
@@ -276,6 +276,23 @@ def save(self, creator, document, based_on_id=None, base_rev=None,
+    class Meta(object):
+        model = DraftRevision
+        fields = ('keywords', 'summary', 'content', 'slug', 'title',
+                  'based_on')
```
Nit: This line doesn't need to wrap.

mythmon commented on kitsune/wiki/forms.py:
```diff
@@ -276,6 +276,23 @@ def save(self, creator, document, based_on_id=None, base_rev=None,
+    def save(self, user, parent_doc, doc_data):
+        """save the draft revision and return the draft revision"""
+        locale = doc_data["locale"]
+        doc_data.pop("locale", None)
```
`.pop` is used to remove *and return* a value from a dict. This is better written as `locale = doc_data.pop('locale')`.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -537,6 +537,22 @@ def edit_document(request, document_slug, revision_id=None):
+    draft_form = DraftRevisionForm(request.POST)
+    if draft_form.is_valid():
+        _create_draft_revision(request, draft_form)
+        return HttpResponse('')
```
Better would be `return HttpResponse(status=204)`. [`HTTP 204`](http://http.cat/204) is "No Content" and is appropriate to use in this context.

safwanrahman replied:
What do you think about returning http 201? as its creating a draft revision, HTTP 201 maybe better status code.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1428,6 +1485,15 @@ def _save_rev_and_notify(rev_form, creator, document, based_on_id=None,
     EditDocumentEvent(new_rev).fire(exclude=new_rev.creator)
 
 
+def _create_draft_revision(request, draft_form):
```
This seems like it should be integrated into `DraftForm` in some way.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -813,23 +829,50 @@ def translate(request, document_slug, revision_id=None):
+    # Check if the user has requested to restore or discard draft
+    draft_request = request.GET.get('form') == 'draft'
+    draft_revision = None
+    # If Draft is available and the user has requested with draft paramaeters
```
Typo: paramaeters should be parameters.

mythmon commented on kitsune/wiki/views.py:
```diff
@@ -1417,6 +1462,18 @@ def _document_form_initial(document):
             'needs_change_comment': document.needs_change_comment}
 
 
+def _draft_initial(type, draft_revision):
```
This seems like it should be two methods, `_draft_initial_doc` and `_draft_initial_rev`.

safwanrahman commented:
@mythmon Updated. r?

brittanystoroz commented:
@safwanrahman I'll take a look through your javascript on this today and let you know how it can be improved.

safwanrahman commented:
Thanks @brittanystoroz. Please let me know so that I can fix

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,20 @@
 
+  var draftbutton = $('.btn-draft');
+  var url = $('.btn-draft' ).data( 'draft-url' );
+  var draftmessage = $('.draft-message').html();
```
When you're declaring a lot of variables all at once, you can just specify `var` once in the beginning, and put each new variable on a new line separated by commas. You can see an example of this [here](https://github.com/mozilla/kitsune/blob/master/kitsune/sumo/static/sumo/js/wiki.js#L125-L131)

It's also a good convention to prefix any variable names with `$` if they represent a jQuery-selected DOM element. So in this block of code you might want to change `draftbutton` to `$draftButton`.

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,20 @@
   }
 
+  var draftbutton = $('.btn-draft');
+  var url = $('.btn-draft' ).data( 'draft-url' );
```
Nit: remove spaces inside the parenthesis in `( 'draft-url'  )`

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,20 @@
+  var draftbutton = $('.btn-draft');
+  var url = $('.btn-draft' ).data( 'draft-url' );
+  var draftmessage = $('.draft-message').html();
+  draftbutton[0].addEventListener('click', function() {
```
jQuery has a built-in `.click` [method we can use](https://api.jquery.com/click/) instead of `addEventListener`. This will remove the need for the `[0]` and clean up the syntax a little bit. So this could be re-written like:

```javascript
$draftButton.click(function() {
   // do things
});
```

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,20 @@
+  draftbutton[0].addEventListener('click', function() {
+    var both_data = $('#both_form').serializeArray();
+    var doc_data = $('#doc_form').serializeArray();
+    var rev_data = $('#rev_form').serializeArray();
```
Nit: camelCase variable names rather than snake_casing. 

Also, can we think of some more meaningful ID names for these form elements? At first glance, it's really hard for me to decipher what `both_form` `doc_form` and `rev_form` mean. If you can help clarify their differences for me a bit I can try to think of some more informative IDs.

safwanrahman replied:
There are 3 type of forms a localizers can see in time by time.
``doc_form`` refers to a form, which contains only document content like title and slug
``rev_form`` refers to a form, which contains revision content like revision content, search summery, and others
``both_form`` refers to a big form, which has both document content and revision content. This form is being showed to the localizers when they localize any article for first time. 
I think the ID name is meaningful as per their use. As we define like the same in the backend views.

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,20 @@
+    posting.done(function() {
+      var time = new Date();
+      var message = draftmessage + time
+      $('.draft-message').css('display', 'block').empty().append(message)
```
Instead of emptying and appending a new message each time, we can simplify this with `.html()`:
`$('.draft-message').html(message);` would replace the entire inner html of the `div` each time. 

For toggling the display of the element, we can use `.show()` instead of `.css()`. We would probably also want this to come *after* we replace the inner content so we're sure the content is updated before we display the element:

`$('.draft-message').html(message).show();`

brittanystoroz commented:
@safwanrahman added some line notes with small improvements to start, couple more things we might want to do here but let's get these in first and take a look again after

safwanrahman commented:
@brittanystoroz updated the javacript. Travis fail seems not related. can you re run the test?

safwanrahman commented:
@brittanystoroz I have updated the patch by adding Error message if saving draft revision fail. also added some CSS for styling the message. What do you think about the fontend  now?

brittanystoroz commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +787,30 @@
     });
   }
 
+  var $draftButton = $('.btn-draft'),
```
Let's take allllll of this javascript you wrote (Lines 790 - 814) and wrap it in a function called `initTranslationDraft` and move it up to line 641 - in between the `initToggleDiff` and `initRevisionList` functions.  Right now it's kind of just hanging out at the bottom of the file, unattached. This will make it a little more consistent with the way the rest of the file is written.

Then let's invoke this new `initTranslationDraft` function on [line 75](https://github.com/safwanrahman/kitsune/blob/draft/kitsune/sumo/static/sumo/js/wiki.js#L75) right beneath `initToggleDiff();`  This will ensure that all of your code is only being used when we're actually on the translation page where it's needed.

safwanrahman commented:
@mythmon r?

brittanystoroz commented:
Travis is finally happy :tada: :tada: :tada:  @mythmon I am good with the front-end on this, wanna take a look at the back-end when you have a chance?

rehandalal commented:
@mythmon and I had a quick chat about this and we think it might be a good idea to refactor this slightly so that you create an `AbstractRevision` class for the common elements of `Revision` and `DraftRevision` and have both those classes inherit from `AbstractRevision`. I can walk you through this if you need help with it.

mythmon commented:
The docs about abstract models are [here](https://docs.djangoproject.com/en/1.7/topics/db/models/#model-inheritance).

mythmon commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -638,6 +639,34 @@
+      })
+        .done(function() {
+          var time = new Date(),
+            message = draftMessage + time;
```
This way of building a message is bad for localization. You should move the text out of the HTML and into the JS and make this line something like `message = k.safeInterpolate(_('Draft has been saved on: {time}'), {time: time});`.

mythmon replied:
This is because not all languages will want to put `{time}` at the end of the string. Some might want to put it in the middle or at the beginning.

mythmon commented on kitsune/wiki/forms.py:
```diff
@@ -276,6 +276,24 @@ def save(self, creator, document, based_on_id=None, base_rev=None,
+        draft, created = DraftRevision.objects.update_or_create(creator=creator,
+                                                                parent_doc=parent_doc,
+                                                                locale=locale,
+                                                                defaults=doc_data)
```
This matches our style guide, but personally I think it is more readable like this:

```python
draft, created = DraftRevision.objects.update_or_create(
    creator=creator, parent_doc=parent_doc, locale=locale, defaults=doc_data)
```

safwanrahman commented:
@mythmon @rehandalal Updated with the Inherited class. r?
I would like to squash it into one before merging. :+1:

mythmon commented on kitsune/wiki/models.py:
```diff
@@ -857,8 +865,6 @@ class Revision(ModelBase, SearchMixin):
     # this is simply the latest of the default locale's revs as of that time.
     # Used to determine whether localizations are out of date.
     based_on = models.ForeignKey('self', null=True, blank=True)
-    # TODO: limit_choices_to={'document__locale':
```
I don't think you have handled this TODO item in this PR. Then again, I don't really know what it is talking about. Do you?

safwanrahman replied:
I think its about having the ``based_on`` into ``settings.WIKI_DEFAULT_LANGUAGE`` only. Lets keep this as it is

rehandalal commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -787,4 +809,5 @@
     });
   }
 
+
```
nit: this seems like an unnecessary extra line

safwanrahman commented:
updated @rehandalal. Is everything Ok?

mythmon commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -1013,6 +1013,143 @@ def test_keywords_require_permission(self):
+        rev = ApprovedRevisionFactory(is_ready_for_localization=True)
+        doc = rev.document
+        doc.is_localizable = True
+        doc.save()
```
You could do this more easily with `rev = ApprovedRevisionFactory(is_ready_for_localization=True, document__is_localizable=True)`.

mythmon commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -1013,6 +1013,143 @@ def test_keywords_require_permission(self):
+            'summary': 'some summary',
+            'title': 'test title',
+            'slug': 'test-slug',
+            'based_on': int(rev.id)}
```
Should `rev.id` already be an int?

safwanrahman replied:
I strongly believed it would be int. but sometime the test fail because it return ``45L`` or something like that.

mythmon commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -1013,6 +1013,143 @@ def test_keywords_require_permission(self):
+        eq_(2, len(trans_content('.user-messages .info form .btn')))
+        # Restore with the draft data
+        draft_request = {'restore': 'Restore'}
+        cache.clear()
```
What's in the cache that you keep clearing?

safwanrahman replied:
``trans_resp = self.client.get(trans_url)``

safwanrahman replied:
at line ``1041``

mythmon replied:
That shouldn't be cached. These tests pass fine for me without clearing the cache. What specifically (what key and value) do you expect to be in the cache?

safwanrahman replied:
I dont know. While requesting the same page in same test, it raised error for me because it was cached. The translation page get cached and when I tried to get the page again, it returned the cached page. 
If test passes without clearing the cache, I am happy to remove it

mythmon commented on kitsune/wiki/tests/test_templates.py:
```diff
@@ -1013,6 +1013,143 @@ def test_keywords_require_permission(self):
+        trans_content = pq(trans_resp.content)
+        eq_(rev1.content, trans_content('.content textarea').text())
+        # As the old based on draft is restored, the latest revision content should not be there
+        assert not rev2.content == trans_content('.content textarea').text()
```
you probably want `assert rev2.content != trans_content('.content textarea').text()`. That's a bit clearer.

safwanrahman replied:
Oh! Yeah.

mythmon commented:
I have some trouble with the tests for this. I think are functional, but they are very muddled. They involve a lot of parts and each test relies on many things. I think the tests should be split up into smaller units, and we need to work on the way you set up the tests.

In particular, you often are using the http interfaces to create models for testing. The only time you should be calling the views (with something like `self.client.get`) is when you are testing the view. Otherwise, you should be creating the models through a factory. You'll need to create a DraftRevisionFactory in `kitsune/wiki/tests/__init__.py`.

I'd go so far as to say that each test should probably only have one http request in it. This is not an absolute rule, it is possible to write a good test that has more than one, but it is probably a good goal to strive for.

Does that make sense? I can give more focused feedback if that would be useful. We can also chat on IRC about this, if it would be easier.

rehandalal commented on kitsune/sumo/static/sumo/less/main.less:
```diff
@@ -307,6 +307,27 @@ input[type="submit"]{
+.alert {
+  border: 1px solid transparent;
+  border-radius: 4px;
+  background: #FFE url("../img/warning-stripes.png") repeat-x scroll left top;
```
nit: alphabetized - ba < bo

mythmon replied:
To clarify: Putting CSS rules in alphabetical order is good style, and actually has a small impact on the compressed file size. Smaller files means faster pages. So it seems silly, but it helps.

safwanrahman replied:
Actually, this is copy from bootstrap. so it was like just be there.

rehandalal commented on kitsune/sumo/static/sumo/less/main.less:
```diff
@@ -307,6 +307,27 @@ input[type="submit"]{
+  margin-bottom: 20px;
+}
+
+.success {
```
same with everything here. alphabetize.

rehandalal commented on kitsune/sumo/static/sumo/less/main.less:
```diff
@@ -307,6 +307,27 @@ input[type="submit"]{
+  border-color: #d6e9c6;
+}
+
+.info {
```
more alphabetizing required.

rehandalal commented on kitsune/wiki/models.py:
```diff
@@ -1055,6 +1063,15 @@ def get_mapping_type(cls):
+class DraftRevision(ModelBase, SearchMixin, AbstractRevision):
+    based_on = models.ForeignKey(Revision, blank=True, null=True)
+    content = models.TextField(blank=True)
+    locale = LocaleField(default=settings.WIKI_DEFAULT_LANGUAGE)
```
do we need this locale field? don't we just use the locale of the document?

safwanrahman replied:
As the document will always be a ``en-US`` one, we can not. We identify with parent document, and locale

rehandalal commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -243,6 +243,21 @@ def setUp(self):
+            'summary': 'some summary',
+            'title': 'test title',
+            'slug': 'test-slug',
+            'based_on': int(based_on.id)
```
should this be an int?

rehandalal replied:
oh @mythmon  already addressed this. it appears all over the tests.

safwanrahman commented:
@mythmon updated the tests. r?

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -470,6 +471,98 @@ def test_needs_change(self):
+        eq_(201, resp.status_code)
+        obj = DraftRevision.objects.get(creator=self.u, document=rev.document, locale='bn-BD')
+        eq_(obj.content, data['content'])
+        eq_(obj.based_on, rev)
```
This test is great. It tests one thing simply and is easy to follow. Nice job!

mythmon commented on kitsune/wiki/tests/test_views.py:
```diff
@@ -470,6 +471,98 @@ def test_needs_change(self):
+        # Get the Translation page of the document
+        trans_url = reverse('wiki.translate', locale=draft.locale, args=[doc.slug])
+        trans_resp = self.client.get(trans_url)
+        eq_(200, trans_resp.status_code)
```
I thinks getting the Translation page of the document is unneeded. The API to store the draft page should be stateless and not depend on the translation page being loaded first. Lines 519 through 521 should probably be removed.

mythmon commented on kitsune/sumo/static/sumo/less/main.less:
```diff
@@ -307,6 +307,27 @@ input[type="submit"]{
+}
+
+.info {
+  background-color: #fcf8e3;
```
This CSS rule is too broad, and is affecting other parts of the site. Specifically, this: 
![shot_14917](https://cloud.githubusercontent.com/assets/305049/12793424/6866b424-ca65-11e5-83a3-aebc5ca86472.png)

@rehandalal any ideas for better CSS rules here? Should they just be scoped to `#localize-document`?

mythmon commented on kitsune/wiki/templates/wiki/translate.html:
```diff
@@ -196,6 +210,12 @@
+            </div>
+            <div id="draft-message-saving" class="alert info" hidden>
+              <img src="/static/sumo/img/customercare/spinner.gif">
+              <strong>{{ _('Draft is saving...........') }} </strong>
```
You don't need 11 `.` here. 3 would do just fine.

mythmon commented on kitsune/sumo/static/sumo/js/wiki.js:
```diff
@@ -638,6 +639,27 @@
+            message = interpolate(gettext('<strong>Draft has been saved on:</strong> %s'), [time]);
+          $('#draft-message-success').html(message).show();
+          $('#draft-message-saving').hide();
+        });
```
You should add a `.fail(function() { ... })` handler to this to show an error message in case the server can't be contacted.

mythmon replied:
The docs here are useful: http://api.jquery.com/jQuery.post/

safwanrahman replied:
added

mythmon commented:
Most of this looks good, and the back end should work well. I'd like to see a couple more improvements to the frontend, and then this should be ready to go.

mythmon commented:
The lint error is fixed on master, so the next time you add commits to this, it should go away.

safwanrahman commented:
@mythmon Updated the front end. r?<
After getting r+, I would squash it into one.

safwanrahman commented:
@mythmon Yap. I like the Idea.
I have made the correction and submitted another commit. can you please check again and let me know?

mythmon commented:
Ok, I think this is ready to go, thank you!

Go ahead and squash this, and then once Travis has passed we can merge it.

safwanrahman commented:
:D Squashed and test passed! :green_apple: 
Its finally ready for merge. :tada: :+1: :100: :zzz:

safwanrahman commented:
@mythmon can merge?

mythmon commented:
Merged. We can deploy this next week.

Osmose commented:
:tada:
